File: lc_soln/1011_capacity_to_ship_packages_within_d_days/solution.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int l = ranges::max(weights);
        int r = reduce(weights.begin(), weights.end());
        int result = 0;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (timeTaken(weights, mid) <= days) {
                result = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return result;
    }

    int timeTaken(const vector<int>& weights, int capacity) {
        int result = 1;
        int curr = 0;
        for (const auto& w : weights) {
            curr += w;
            if (curr > capacity) {
                curr = w;
                ++result;
            }
        }
        return result;
    }
};

int main() {
    Solution solution;
    vector<int> weights = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    cout << solution.shipWithinDays(weights, 5) << "\n";
    return 0;
}

File: lc_soln/1011_capacity_to_ship_packages_within_d_days/solution.py
from typing import List


class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        def time_taken(weights, capacity):
            result = 1
            curr = 0
            for w in weights:
                curr += w
                if curr > capacity:
                    curr = w
                    result += 1
            return result

        l = max(weights)
        r = sum(weights)

        result = -1
        while l <= r:
            mid = l + (r - l) // 2
            if time_taken(weights, mid) <= days:
                result = mid
                r = mid - 1
            else:
                l = mid + 1
        return result
